app/esper/rekall_vgrid.py

```diff
@@ -0,0 +1,354 @@
+each track shows a timeline of intervals in that track, as well as drawing
+objects and showing metadata of the interval as configured.
+
+There are two main use cases of VGrid:
+    1. VideoVBlocks. This mode displays a VBlock for each video. Each track
+        in the VBlock is an IntervalSet to visualize in that video.
+    2. IntervalVBlocks. This mode displays a VBlock for each interval/clip.
+        Each track in the VBlock is an IntervalSet to visualize in that
+        interval/clip.
+
+The VBlocksBuilders in this module helps building the JSON structure for this
+list of VBlocks.
+
+Example of using VideoVBlock:
+
+    # face_collection is a DomainIntervalCollection of face intervals.
+
+    # We want to see a list of VBlocks where each is a video in the collection
+    # We want to see two tracks in each VBlock: one showing all faces in the
+    # collection with bounding boxes drawn and a flag metadata set; another
+    # showing all captions of the video.
+
+    json = VideoVBlocksBuilder()\\
+        .add_track(
+            VideoTrackBuilder('faces', face_collection)\\
+                .set_draw_type(DrawType_Bbox())\\
+                .add_metadata('flag', Metadata_Flag()))\\
+        .add_track(
+            CaptionTrack())\\
+        .build()
+
+Example of using IntervalBlock:
+
+    # conversations_with_ids is a DomainIntervalCollection of conversation
+    # sequences, where the payload on each conversation is an IntervalSet of
+    # face bounding boxes with identity as payload.
+
+    # We want to see a list of VBlocks where each is a conversation 
+    # sequence in the collection. We want to see two tracks in each VBlock:
+    # one showing the conversation sequence; another showing all the faces with
+    # the bounding boxes drawn and generic metadata showing the identities.
+
+    json = IntervalVBlocksBuilder()\\
+        .add_track(
+            IntervalTrackBuilder('conversation'))\\
+        .add_track(
+            IntervalTrackBuilder('faces', lambda i:i.payload)\\
+                    .set_draw_type(DrawType_Bbox())\\
+                    .add_metadata('identity', Metadata_Generic()))\\
+        .build(conversations_with_ids)
+
+Symbols exported in this module are the following.
+
+VBlocksBuilders:
+    VideoVBlocksBuilder: VideoVBlocks mode.
+    IntervalVBlocksBuilder: IntervalVBlocks mode.
+
+Tracks:
+    VideoTrackBuilder: Build custom tracks for VideoVBlocks mode.
+    IntervalTrackBuilder: Build custom tracks for IntervalVBlocks mode.
+    CaptionTrack: Add caption to the VBlocks. Works in either mode.
+
+DrawTypes:
+    DrawType_Bbox: Draw bounding boxes around the spatial dimension.
+    DrawType_Caption: Display text in the caption box.
+
+Metadatas:
+    Metadata_Flag: Plant a flag marking the temporal interval.
+    Metadata_Generic: Show a generic JSON text.
+"""
+
+from rekall.interval_set_3d import IntervalSet3D, Interval3D
```

Comment:
Oh one other thing. Let's not put this in Esper, let's either keep this in Rekall or put it in a new module. This should be usable outside of Esper

Reply:
The way I understand it, we have VGrid and Rekall as two self-contained libraries; Esper is a framework that defines some default ways to hook up a database, VGrid and Rekall and the jupyter environment together, either locally or in google cloud. Therefore I think this module should live in Esper. Thoughts?

Reply:
That's a reasonable way to think about it. I'm still semi-opposed to the notion of VGrid and Rekall as being self-contained in that both rely on a fundamental notion of intervals and interval algebra, and that abstraction should be used consistently across our codebase. Ideally, Rekall would actually define an equivalent Javascript interface for interval sets and VGrid would depend on Rekall for that API.

But either way, Esper right now is best thought of as "end-to-end data science workbench, bells and whistles included." However, VGrid and Rekall should be able to work outside of Esper, e.g. I'm preparing a demo for Strata and it's convenient to just make a one-off Jupyter notebook with a little VGrid/Rekall code than a whole new Esper instance. However, for that to be useful, I need Rekall/VGrid glue that isn't tied to Esper.